### PR TITLE
Add webhooks and badge callback

### LINKED ISSUE #62: Add trigger to call web hooks for resources
Implement web hooks so data in Razeedash can be augmented by third-party services such as test suites and vulnerability scanners.

Components are:

- Web hook creation
- Web hook deletion
- Trigger by resource ID
- Trigger by Regex
- Callback response

### Web hook definition

`POST /v2/webhook/` will create a web hook and the header should contain the `razee-org-key`

JSON body for web hook triggered by cluster change:

```json
{
          "cluster_id": "ID of the cluster",
          "trigger": "cluster",
          "kind": "Deployment",
          "field": "searchableData.name",
          "filter": "regex string example, `(watch-keeper)`",
          "service_url": "URL of service to POST upon triggering"
}
```

**Note: field and filter are optional. If not defined the above will fire if a `deployment` resource kind is changed**

JSON body for web hook triggered by image change:

```json
{
          "trigger": "image",
          "kind": "image",
          "field": "name",
          "filter": "regex string example, `(quay.io\\/mynamespace)`",
          "service_url": "URL of service to POST upon triggering"
}
```

**Note: field and filter are optional. If not defined the above will fire if any new image is deployed on an organization's clusters**

- kind: Kind would be the type of resource or in a special case, images
- trigger: cluster, image
- id: ID of the resource used in trigger OR
- field: field dot-notation into JSON to apply a filter
- filter: regex parameter to filter if the trigger should fire or not if field is defined
- service URL:  URL to call if the web hook is triggered

### Web hook deletion

`DELETE /v2/webhook/:id` will delete the web hook.   Current badge data will not be affected by the deletion.   If a callback for that webhook_id  occurs then the badge would be removed from the resource and 404 sent back to the calling service. 

### Trigger Logic

Trigger points will be added to Razeedash API for when:
- New image is deployed
- A resource kind is deployed or modified on a specific cluster

Trigger will also have to pass regex filter in order to fire the web hook.  If determine to call the web hook the service URL is called and relevant data posted along with a callback URL.   The called service can augment the data by calling the callback URL.

Razeedash API will look up the resource to make sure it is still being used and if not, return a 404 indicating to the calling service it should no longer provide updates on this web hook.   If found the badge information will added to resource.

Calling the service by cluster trigger POSTs:

```json
{
  "cluster_id": "ID of the cluster",
  "cluster_metadata": "JSON array for the cluster metadata",
  "resource_id": "ID of the resource to badge",
  "resource_kind": "What kind of resource",
  "resource": "JSON of the resource object",
  "webhook_id": "ID of the webhook definition",
  "callback_url":  "URL to POST badge data"
}
```

Calling the service by image trigger POSTs:

```json
{
  "image_id": "ID of the image",
  "image_name": "Image name",
  "webhook_id": "ID of the webhook definition",
  "callback_url":  "URL to POST badge data"
}
```
**Scenario: Trigger by resource ID**

User defines a web hook that if a deployment changes or something new is added to call a web hook to run integration tests.

```json
{
  "kind": "Deployment",
  "trigger": "cluster",
  "cluster_id": "fb56c61b676844d292f1f18e719c31f2",
  "service_url": "https://my.testingservice.com/run_integration"
}
```

New deployment is rolled out to cluster designated as the "staging" cluster environment.  Web hook is called to service_url and the deployment.yaml, org_id, cluster_id are all posted along with a a call back URL. 

The testing service calls the callback with badge data:

- badge: URL of a running man
- description: "Running tests"
- link: (link to the live tests being run)
- status: info

Testing service runs integration tests on the staging cluster and and calls the callback with badge data:

- badge: URL of a green circle
- description: "All tests completed successfully"
- link: (link to the test logs)
- status: info

**Scenario:  Trigger by regex**

User defines a web hook that if a deployment changes or something new is added to call a web hook to run integration tests.

```json
{
  "kind": "image",
  "trigger": "image",
  "field": "name",
  "filter": "(quay.io\/mynamespace)",
  "service_url": "https://my.quayscanner.com/check"
}
```
When a new image is deployed, the name of the image is checked against the filter, if defined, and then the service_url is called with image name, image ID, org_id and callback URL.

The scanner service calls the callback with badge data:

- badge: URL image of binoculars
- description: "Looking for vulnerabilities"
- link: (link to service of that image being checked)
- status: info

Razeedash API checks to see if the image is still in use and returns a 201

The image is scanned and shown clean and will be rechecked by the service in 24 hours.  In the meantime it calls the callback with badge data:

- badge: URL image of green circle
- description: "no vulnerabilities"
- link: (link to service of that image results)
- status: info

Razeedash API checks to see if the image is still in use and returns a 201

24 hours late a minor vulnerabilities is discovered and the service calls the callback again with badge data:

- badge: URL image of yellow circle
- description: "minor vulnerabilities detected"
- link: (link to service of that image results)
- status: warning

Razeedash API checks to see if the image is still in use and returns a 201

24 hours later a major vulnerability is discovered and the service calls the callback again with badge data:

- badge: URL image of red circle
- description: "major vulnerabilities detected"
- link: (link to service of that image results)
- status: error

Razeedash API checks to see if the image is still in use and finds it is not and returns a 404 to the vulnerability service.  The vulnerability service should then stop reporting that image security issues from now on.

**Scenario: Filter a specific resource**

User defines a web hook that filters for a specific resource.  In this case we are looking for new or changed deployments where the field `metadata.name` matches `watch-keeper` on a specific cluster to trigger the web hook.

```json
{
  "kind": "Deployment",
  "field": "metadata.name",
  "filter": "(watch-keeper)",
  "trigger": "cluster",
  "id": "fb56c61b676844d292f1f18e719c31f2",
  "service_url": "https://my.testingservice.com/run_integration"
}
```

New deployment is rolled out to cluster designated as the "staging" cluster environment.  Web hook is called to service_url and the deployment.yaml, org_id, cluster_id are all posted along with a a call back URL. 

### Callback response from remote service

When the remote service wants to add a badge as a result of the web hook call, the POST to /v2/callback should have the following body:

Header should contain the `razee-org-key`
```json
{
    "webhook_id": "id of the webhook from initial call to service",
    "url": "URL of the badge image",
    "description": "short description of badge",
    "link": "URL link for details",
    "status": "info | error | warning"
}
```

Razeedash API will accept the callback URL and make sure the webhook and resource is still valid.  If the resource is no longer in use or the webhook was deleted then the callback response will return a 404.  If valid it will add the augmented data to the resource.

The resource will have a new attribute `badges`.  The badge will replace any existing badge with the same webhook_id or if it does not exist, add to the array.

### PR DESCRIPTION
**Pull request by geoffcorey**
Resolves #62

### CODE REVIEW

**Comment by geoffcorey** [comment]
```
----------------------------------|----------|----------|----------|----------|-------------------|
File                              |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
----------------------------------|----------|----------|----------|----------|-------------------|
All files                         |    47.48 |    52.86 |    32.94 |    47.73 |                   |
 razeedash-api                    |        0 |        0 |        0 |        0 |                   |
  gwc.js                          |        0 |      100 |        0 |        0 |... 34,35,36,38,42 |
  gwc1.js                         |        0 |        0 |        0 |        0 |... 09,413,416,417 |
  mm.js                           |        0 |      100 |        0 |        0 |... 33,34,35,37,40 |
 razeedash-api/app                |     4.23 |     7.41 |        0 |     4.23 |                   |
  conf.js                         |        0 |        0 |      100 |        0 |             17,30 |
  index.js                        |        0 |        0 |        0 |        0 |... 90,92,93,94,96 |
  log.js                          |      100 |      100 |      100 |      100 |                   |
  wait-mongo.js                   |        0 |      100 |        0 |        0 |... 36,37,42,43,44 |
 razeedash-api/app/mongo          |        0 |        0 |        0 |        0 |                   |
  mongoClient.js                  |        0 |        0 |        0 |        0 |... 35,136,137,139 |
 razeedash-api/app/routes         |        0 |        0 |        0 |        0 |                   |
  index.js                        |        0 |        0 |        0 |        0 |... 76,142,143,146 |
 razeedash-api/app/routes/install |        0 |      100 |        0 |        0 |                   |
  index.js                        |        0 |      100 |        0 |        0 |... 04,106,107,112 |
 razeedash-api/app/routes/kube    |        0 |      100 |        0 |        0 |                   |
  kube.js                         |        0 |      100 |        0 |        0 |    17,18,21,22,25 |
 razeedash-api/app/routes/v2      |    95.03 |    81.03 |      100 |    95.03 |                   |
  clusters.js                     |    97.44 |    78.38 |      100 |    97.44 |       112,116,118 |
  webhooks.js                     |    90.63 |    85.71 |      100 |    90.63 | 53,54,61,62,65,66 |
 razeedash-api/app/s3             |        0 |        0 |        0 |        0 |                   |
  s3Client.js                     |        0 |        0 |        0 |        0 |... 96,101,102,106 |
 razeedash-api/app/utils          |    95.31 |    82.67 |     87.5 |    96.32 |                   |
  bunyan.js                       |    91.67 |    91.67 |      100 |    91.67 |                25 |
  cluster.js                      |    94.37 |    81.08 |       90 |    94.37 |     30,71,103,118 |
  crypt.js                        |      100 |      100 |      100 |      100 |                   |
  orgs.js                         |      100 |      100 |      100 |      100 |                   |
  webhook.js                      |    95.45 |    77.27 |    77.78 |    97.67 |           145,162 |
----------------------------------|----------|----------|----------|----------|-------------------|
```

File: README.md [hunk 1]
```diff
@@ -206,3 +206,247 @@ Check logs across pods using `kc_logs.sh` script from
...
+
+- Web hook creation
+- Web hook deletion
+- Trigger by resource ID
+- Trigger by Regex
+- Callback response
+
+### Web hook definition
+
+`POST /v2/webhook/` will create a web hook and the header should contain the `razee-org-key`
```
Comment (adamkingit) [inline README.md]: should -> must?

File: README.md [hunk 2]
```diff
@@ -206,3 +206,247 @@ Check logs across pods using `kc_logs.sh` script from
...
+```
+
+**Note:** field and filter are optional. If not defined the above will fire if
+any new image is deployed on an organization's clusters
+
+- kind: Kind would be the type of resource or in a special case, images
+- trigger: cluster, image
+- id: ID of the resource used in trigger OR
+- field: field dot-notation into JSON to apply a filter
+- filter: regex parameter to filter if the trigger should fire or not if field
```
Comment (adamkingit) [inline README.md]: Does webhook trigger whenever field matches the filter, or does filter omit any change where field matches filter?

File: README.md [hunk 3]
```diff
@@ -206,3 +206,247 @@ Check logs across pods using `kc_logs.sh` script from
...
+Components are:
+
+- Web hook creation
+- Web hook deletion
+- Trigger by resource ID
+- Trigger by Regex
+- Callback response
+
+### Web hook definition
+
```
Comment (adamkingit) [inline README.md]: Do we have or need a `GET` to allow the UI to get the registered webhooks for management?

File: README.md [hunk 4]
```diff
@@ -206,3 +206,247 @@ Check logs across pods using `kc_logs.sh` script from
...
+hours.  In the meantime it calls the callback with badge data:
+
+- badge: URL image of green circle
+- description: "no vulnerabilities"
+- link: (link to service of that image results)
+- status: info
+
+Razeedash API checks to see if the image is still in use and returns a 201
+
+24 hours late a minor vulnerabilities is discovered and the service calls the
```
Comment (adamkingit) [inline README.md]: later

File: README.md (hunk 1 above)
Comment (geoffcorey) [reply README.md]: done

File: README.md (hunk 2 above)
Comment (geoffcorey) [reply README.md]: Matches

File: README.md (hunk 4 above)
Comment (geoffcorey) [reply README.md]: done

File: README.md (hunk 3 above)
Comment (geoffcorey) [reply README.md]: We could add it for completeness but I would imagine the UI will directly access mongoDB.

File: README.md (hunk 2 above)
Comment (geoffcorey) [reply README.md]: Updated docs

File: app/routes/v2/clusters.js [hunk 5]
```diff
@@ -137,7 +149,7 @@ const updateClusterResources = async (req, res, next) => {
                 ...pushCmd
               },
               { upsert: true }
-            );
+            );
```
Comment (adamkingit) [inline app/routes/v2/clusters.js]: delete trailing whitespace

File: app/routes/v2/webhooks.js [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+      } else {
+        // Validate badge properties
+        let badge = req.body;
+        badge.webhook_id = req.params.webhook_id;
+        let properties = ['webhook_id', 'url', 'description', 'link', 'status'];
+        let isValid = jkValidate(badge, properties);
+        if ((webhook.trigger == WEBHOOK_TRIGGER_IMAGE) && (!badge.image_id)) {
+          properties.push('image_id');
+          isValid = false;
+        }
```
Comment (adamkingit) [inline app/routes/v2/webhooks.js]: Suggest alternate implementation
``` javascript 
        if (webhook.trigger == WEBHOOK_TRIGGER_IMAGE) {
          properties.push('image_id');
        }
        let isValid = jkValidate(badge, properties);

```

File: app/routes/v2/webhooks.js [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+};
+
+const addWebhook = async (req, res, next) => {
+  try {
+    let webhook = req.body;
+    const Webhooks = req.db.collection('webhooks');
+    if (webhook.trigger == WEBHOOK_TRIGGER_CLUSTER) {
+      const Clusters = req.db.collection('clusters');
+      const result = await Clusters.findOne({ cluster_id: webhook.cluster_id, org_id: req.org._id });
+      if ((!result) || (result && result.deleted == true)) {
```
Comment (adamkingit) [inline app/routes/v2/webhooks.js]: Will this prevent a deleted webhook from being added again?

Comment (adamkingit) [reply app/routes/v2/webhooks.js]: nm, this is a deleted cluster

File: app/routes/v2/webhooks.js [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+        res.status(404).send(`Cluster ${webhook.cluster_id} not found or has been deleted`);
+      } else {
+        await Webhooks.insertOne(webhook);
```
Comment (adamkingit) [inline app/routes/v2/webhooks.js]: We should add the ord_id to the webhook document to facilitate future cleanup when an org is deleted.

File: app/routes/v2/webhooks.js (hunk 6 above)
Comment (geoffcorey) [reply app/routes/v2/webhooks.js]: changed

File: app/routes/v2/webhooks.js (hunk 7 above)
Comment (geoffcorey) [reply app/routes/v2/webhooks.js]: This makes sure the cluster exists and is not deleted for a webhook cluster trigger that you are trying to add

File: app/routes/v2/webhooks.js (hunk 8 above)
Comment (geoffcorey) [reply app/routes/v2/webhooks.js]: done

File: app/utils/webhook.js [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+      var field = objectPath.get(resourceObj, webhook.field);
+      if (!field.match(webhook.filter)) {
+        match = false;
+      }
+    }
+    if (match) {
+      postData.webhook_id = webhook._id;
+      if (!await fireWebhook(webhook, postData, req)) {
+        success = false;
+      }
```
Comment (adamkingit) [inline app/utils/webhook.js]: Alternate impl:
``` javascript 
success = !await fireWebhook(webhook, postData, req);
```
Do we want the ! of fireWebhook(), or do we want to send its result along?

File: app/utils/webhook.js [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+    const callbackURL = new URL('v2/callback', process.env.RAZEEDASH_API_URL);
+    const Webhooks = req.db.collection('webhooks');
+    const webhooks = await Webhooks.find({ org_id: req.org._id, trigger: WEBHOOK_TRIGGER_IMAGE }).toArray();
+    const postData = {
+      org_id: req.org._id,
+      image_name: name,
+      image_id: image_id,
+      callback_url: callbackURL
+    };
+    return processWebhooks(webhooks, postData, { name: name, image_id: image_id }, req);
```
Comment (adamkingit) [inline app/utils/webhook.js]: I think we have to await in order to have a chance at catching an exception. 
```javascript 
return await processWebhooks(...
```

File: app/utils/webhook.js [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+    const postData = {
+      org_id: req.org._id,
+      cluster_id: clusterId,
+      cluster_metadata: metadata,
+      resource_id: clusterId,
+      resource_kind: resourceObj.searchableData.kind,
+      resource: resourceObj,
+      callback_url: callbackURL
+    };
+    return processWebhooks(webhooks, postData, resourceObj, req);
```
Comment (adamkingit) [inline app/utils/webhook.js]: also await here?

File: app/utils/webhook.js (hunk 9 above)
Comment (geoffcorey) [reply app/utils/webhook.js]: changed to success = await fireWebhook.   we want to know if it failed.

File: app/utils/webhook.js (hunk 10 above)
Comment (geoffcorey) [reply app/utils/webhook.js]: done

File: app/utils/webhook.js (hunk 11 above)
Comment (geoffcorey) [reply app/utils/webhook.js]: done

File: app/utils/webhook.js (hunk 9 above)
Comment (geoffcorey) [reply app/utils/webhook.js]: done

**Pull request merged**